src/main/starlark/builtins_bzl/common/cc/experimental_cc_static_library.bzl

```diff
@@ -0,0 +1,301 @@
+# Copyright 2021 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This is an experimental implementation of cc_static_library.
+
+We may change the implementation at any moment or even delete this file. Do not
+rely on this.
+"""
+
+load(":common/cc/action_names.bzl", "ACTION_NAMES")
+load(":common/cc/cc_common.bzl", "cc_common")
+load(":common/cc/cc_helper.bzl", "artifact_category", "cc_helper")
+load(":common/cc/cc_info.bzl", "CcInfo")
+load(":common/cc/semantics.bzl", "semantics")
+
+cc_internal = _builtins.internal.cc_internal
+
+def _declare_static_library(*, name, actions, cc_toolchain):
+    new_name = cc_internal.get_artifact_name_for_category(
+        cc_toolchain = cc_toolchain,
+        category = artifact_category.STATIC_LIBRARY,
+        output_name = cc_helper.get_base_name(name),
+    )
+    return actions.declare_file(cc_helper.replace_name(name, new_name))
+
+def _collect_linker_inputs(deps):
+    transitive_linker_inputs = [dep[CcInfo].linking_context.linker_inputs for dep in deps]
+    return depset(transitive = transitive_linker_inputs)
+
+def _flatten_and_get_objects(linker_inputs):
+    # Flattening a depset to get the action inputs.
+    transitive_objects = []
+    for linker_input in linker_inputs.to_list():
+        for lib in linker_input.libraries:
+            if lib.pic_objects:
```

Comment:
Should be `if lib.pic_objects != None`. There are non-empty archives in C++ linking code, which have objects set to `None`. And empty ones that have objects set to `[]`.